allennlp/semparse/contexts/text2sql_table_context.py

```diff
@@ -0,0 +1,142 @@
+GRAMMAR_DICTIONARY["source_table"] = ['table_name ws "AS" wsp name']
+GRAMMAR_DICTIONARY["source_subq"] = ['("(" ws query ws ")" ws "AS" ws name)', '("(" ws query ws ")")']
+
+GRAMMAR_DICTIONARY["where_clause"] = ['(ws "WHERE" wsp expr where_conj)', '(ws "WHERE" wsp expr)']
+GRAMMAR_DICTIONARY["where_conj"] = ['(ws "AND" wsp expr where_conj)', '(ws "AND" wsp expr)']
+
+GRAMMAR_DICTIONARY["groupby_clause"] = ['(ws "GROUP" ws "BY" group_clause ws "HAVING" ws expr)',
+                                        '(ws "GROUP" ws "BY" group_clause)']
+GRAMMAR_DICTIONARY["group_clause"] = ['(ws expr ws "," group_clause)', '(ws expr)']
+
+GRAMMAR_DICTIONARY["orderby_clause"] = ['ws "ORDER" ws "BY" order_clause']
+GRAMMAR_DICTIONARY["order_clause"] = ['(ordering_term ws "," order_clause)', 'ordering_term']
+GRAMMAR_DICTIONARY["ordering_term"] = ['(ws expr ordering)', '(ws expr)']
+GRAMMAR_DICTIONARY["ordering"] = ['(ws "ASC")', '(ws "DESC")']
+GRAMMAR_DICTIONARY["limit"] = ['ws "LIMIT" ws number']
+
+GRAMMAR_DICTIONARY["col_ref"] = ['(table_name ws "." ws column_name)', 'column_name']
+GRAMMAR_DICTIONARY["table_name"] = ['name']
+GRAMMAR_DICTIONARY["column_name"] = ['name']
+GRAMMAR_DICTIONARY["ws"] = ['~"\s*"i']
+GRAMMAR_DICTIONARY['wsp'] = ['~"\s+"i']
+GRAMMAR_DICTIONARY['name'] = ['~"[a-zA-Z]\w*"i']
+
+GRAMMAR_DICTIONARY["expr"] = ['in_expr', 'like_expr', 'between_expr', 'binary_expr',
+                              'unary_expr', 'null_check_expr', 'source_subq', 'value']
+GRAMMAR_DICTIONARY["like_expr"] = ['value wsp "LIKE" wsp string']
+GRAMMAR_DICTIONARY["in_expr"] = ['(value wsp "NOT" wsp "IN" wsp string_set)',
+                                 '(value wsp "IN" wsp string_set)',
+                                 '(value wsp "NOT" wsp "IN" wsp expr)',
+                                 '(value wsp "IN" wsp expr)']
+
+GRAMMAR_DICTIONARY["between_expr"] = ['value ws "BETWEEN" wsp value ws "AND" wsp value']
+GRAMMAR_DICTIONARY["binary_expr"] = ['value ws binaryop wsp expr']
+GRAMMAR_DICTIONARY["unary_expr"] = ['unaryop expr']
+GRAMMAR_DICTIONARY["null_check_expr"] = ['(col_ref ws "IS" ws "NOT" ws "NULL")', '(col_ref ws "IS" ws "NULL")']
+
+GRAMMAR_DICTIONARY["value"] = ['parenval', '"YEAR(CURDATE())"', 'number', 'boolean',
+                               'function', 'col_ref', 'string']
+GRAMMAR_DICTIONARY["parenval"] = ['"(" ws expr ws ")"']
+GRAMMAR_DICTIONARY["function"] = ['(fname ws "(" ws "DISTINCT" ws arg_list_or_star ws ")")',
+                                  '(fname ws "(" ws arg_list_or_star ws ")")']
+
+GRAMMAR_DICTIONARY["arg_list_or_star"] = ['arg_list', '"*"']
+GRAMMAR_DICTIONARY["arg_list"] = ['(expr ws "," ws arg_list)', 'expr']
+GRAMMAR_DICTIONARY["number"] = ['~"\d*\.?\d+"i']
+GRAMMAR_DICTIONARY["string_set"] = ['ws "(" ws string_set_vals ws ")"']
+GRAMMAR_DICTIONARY["string_set_vals"] = ['(string ws "," ws string_set_vals)', 'string']
+GRAMMAR_DICTIONARY["string"] = ['~"\'.*?\'"i']
```

Comment:
We will have to update this not to use regexes to actually get the SQL queries because right now this will give us an action that looks like `string -> ['~"\'.*?\'"i']` that represents a variable. If there are multiple variables then we won't be able to match them to the correct annotated values. For example, we will get `string -> ['~"\'.*?\'"i']` in the action sequence twice for the query below and we won't know which is the actor and which is the director.

```
        "sql": [
            "SELECT COUNT( DISTINCT ( ACTORalias0.NAME ) ) FROM ACTOR AS ACTORalias0 , CAST AS CASTalias0 , DIRECTED_BY AS DIRECTED_BYalias0 , DIRECTOR AS DIRECTORalias0 , MOVIE AS MOVIEalias0 WHERE ACTORalias0.NATIONALITY = \"actor_nationality0\" AND CASTalias0.AID = ACTORalias0.AID AND DIRECTORalias0.DID = DIRECTED_BYalias0.DID AND DIRECTORalias0.NAME = \"director_name0\" AND MOVIEalias0.MID = CASTalias0.MSID AND MOVIEalias0.MID = DIRECTED_BYalias0.MSID ;"
        ],
        "variables": [
            {
                "example": "Iran",
                "location": "both",
                "name": "actor_nationality0",
                "type": "actor_nationality"
            },
            {
                "example": "Jim Jarmusch",
                "location": "both",
                "name": "director_name0",
                "type": "director_name"
            }
        ]
```
I think what we want is `GRAMMAR_DICTIONARY["string"] = [\"director_name0\",  "actor_nationality0"]`, which we can just get from the name field, so when we postprocess the action sequence we can put director name and action nationality in the right slots.

Reply:
I think the same thing applies for the `number` nonterminal in the grammar.

Reply:
Oh yep good point - i'll do that in a separate PR.